### PR TITLE
[Auto Parallel]: Support optional, inplace input and output for DistTensor.

### PR DESCRIPTION
**Pull request by GhostScreaming**
### PR types
Others

### PR changes
Others

### Description
Pcard-73145

1. Support `paddle::optional<phi::Tensor>` and `paddle::optional<std::vector<phi::Tensor>>` input and output for `DistTensor`. `Expand_As`, `Bincount`, `Linear_Interp` forward and backward are verified. 
2. Support inplace output. Including following `inplace phi::Tensor`, `inplace std::vector<phi::Tensor>`, `inplace paddle::optional<phi::Tensor>` and `inplace paddle::optional<std::vector<phi::Tensor>>` type of outputs. `check_finite_and_unscale_`, `adamax_` and `merged_adam_` are verified.
3. Operators have output of `std::tuple<...>` are supported now as all type of output are supported.

### CODE REVIEW

File: paddle/phi/api/yaml/generator/dist_api_gen.py [hunk 1]
```diff
@@ -123,8 +123,26 @@
       {name}_meta_ptr_vec[i] = &{name}_meta_vec[i];
     }}
 """
+
 # TODO(GhostScreaming): support optional args later
```
Comment (chenwhql) [inline paddle/phi/api/yaml/generator/dist_api_gen.py]: 这个TODO是不是可以移除了

File: paddle/phi/api/yaml/generator/dist_api_gen.py [hunk 2]
```diff
@@ -123,8 +123,26 @@
...
-OPTIONAL_GLOBAL_VECTOR_META_IN_TEMPLATE = """
+OPTIONAL_GLOBAL_SINGLE_META_IN_TEMPLATE = """meta_dist_{}, """
+OPTIONAL_GLOBAL_SINGLE_META_IN_DECL_TEMPLATE = """
+    phi::MetaTensor meta_dist_{name} = {name} ? MakeMetaTensor(*(*{name}).impl()) : phi::MetaTensor();
```
Comment (chenwhql) [inline paddle/phi/api/yaml/generator/dist_api_gen.py:132]: 其实我们现在做了两遍meta tensor的构造，前面MakeDistMetaTensor也构造了一遍，后面我们可以看下能否在下面复用前面MakeDistMetaTensor的结果，从而降低调度开销

File: paddle/phi/api/yaml/generator/dist_api_gen.py [hunk 3]
```diff
@@ -123,8 +123,26 @@
...
+OPTIONAL_GLOBAL_VECTOR_META_IN_TEMPLATE = """{}_meta_ptr_vec, """
+OPTIONAL_GLOBAL_VECTOR_META_IN_DECL_TEMPLATE = """
+    std::vector<phi::MetaTensor> {name}_meta_vec_tmp;
+    if ({name}) {{
+      for (auto tmp : *{name}) {{
+        {name}_meta_vec_tmp.emplace_back(MakeMetaTensor(*tmp.impl()));
+      }}
+    }}
+    std::vector<const phi::MetaTensor*> {name}_meta_ptr_vec_tmp({name}_meta_vec_tmp.size());
+    for (size_t i=0; i<{name}_meta_ptr_vec_tmp.size(); i++) {{
```
Comment (chenwhql) [inline paddle/phi/api/yaml/generator/dist_api_gen.py]: `size_t i = 0; i < {name}_meta_ptr_vec_tmp.size(); ++i`
循环条件建议加空格，一般用++i

File: test/auto_parallel/test_dist_tensor.py [hunk 4]
```diff
@@ -138,6 +141,107 @@ def test_unbind_api_for_dist_tensor(self):
...
+        weight = np.random.random(size=[16]).astype("float32")
+        local_x, dist_x = self.create_local_and_dist_tensor_pair(x)
+        local_weight, dist_weight = self.create_local_and_dist_tensor_pair(weight)
+
+        local_out = paddle.bincount(local_x, weights=local_weight)
+        dist_out = paddle.bincount(dist_x, weights=dist_weight)
+
+        self.check_tensor_eq(local_out, dist_out)
+
+    # def test_layer_norm_api_for_dist_tensor(self):
```
Comment (chenwhql) [inline test/auto_parallel/test_dist_tensor.py]: 如果后面需要但暂时没跑通，建议先备份在本地，大段注释先删除

File: test/auto_parallel/test_dist_tensor.py [hunk 5]
```diff
@@ -138,6 +141,107 @@ def test_unbind_api_for_dist_tensor(self):
...
+            -1, # out_d
+            -1, # out_h
+            50, # in_w * out_w
+            scale,
+            'linear', # interp_method
+            False, # align_corners
+            1, # align_mode
+        )
+        # print("local_out: ", local_out)
+        dist_out = paddle._C_ops.linear_interp(
```
Comment (chenwhql) [inline test/auto_parallel/test_dist_tensor.py:236]: 这个用正式api paddle.nn.functional.interpolate 可以测试吗

File: paddle/phi/api/yaml/generator/dist_api_gen.py (hunk 1 above)
Comment (GhostScreaming) [reply paddle/phi/api/yaml/generator/dist_api_gen.py]: Done, thx.

File: paddle/phi/api/yaml/generator/dist_api_gen.py (hunk 2 above)
Comment (GhostScreaming) [reply paddle/phi/api/yaml/generator/dist_api_gen.py:132]: 收到~

File: paddle/phi/api/yaml/generator/dist_api_gen.py (hunk 3 above)
Comment (GhostScreaming) [reply paddle/phi/api/yaml/generator/dist_api_gen.py]: Done, thx.

File: test/auto_parallel/test_dist_tensor.py (hunk 4 above)
Comment (GhostScreaming) [reply test/auto_parallel/test_dist_tensor.py]: Done, thx.

File: test/auto_parallel/test_dist_tensor.py (hunk 5 above)
Comment (GhostScreaming) [reply test/auto_parallel/test_dist_tensor.py:236]: 需要测试`linear_interp`的`paddle::optional<phi::Tensor> out_size`和`paddle::optional<std::vector<phi::Tensor>> size_tensor`类型的输入，看了一下`paddle.nn.functional.interpolate`的代码实现，`size_tensor`是传不进去的。

File: paddle/phi/core/distributed/auto_parallel/dist_tensor.h [hunk 6]
```diff
@@ -58,6 +58,10 @@ class DistTensor final
   /// \return void
   void set_dims(const DDim& dims);
 
+  /// \brief Set the global dims of the dist tensor after kernel execution.
+  /// \return void
+  void unsafe_set_dims(const DDim& dims);
```
Comment (LiYuRio) [inline paddle/phi/core/distributed/auto_parallel/dist_tensor.h:59]: 上面的那个set_dims应该删掉吧

File: paddle/phi/api/lib/api_gen_utils.cc [hunk 7]
```diff
@@ -590,5 +590,43 @@ std::vector<phi::distributed::DistTensor*> SetKernelDistOutput(
...
+  if (out) {
+    results = std::vector<phi::DenseTensor*>(out->size(), nullptr);
+    for (size_t i = 0; i < out->size(); ++i) {
+      results[i] = static_cast<phi::DenseTensor*>(out->at(i).impl().get());
+    }
+  }
+  return results;
+}
+
+*/
```
Comment (LiYuRio) [inline paddle/phi/api/lib/api_gen_utils.cc]: 这里应该删掉

Comment (GhostScreaming) [reply paddle/phi/api/lib/api_gen_utils.cc]: Done, thx.

File: paddle/phi/core/distributed/auto_parallel/dist_tensor.h (hunk 6 above)
Comment (GhostScreaming) [reply paddle/phi/core/distributed/auto_parallel/dist_tensor.h:59]: Done, thx.

**Review by chenwhql — APPROVED** [review]
LGTM

**Pull request merged**